### PR TITLE
Fix #191, fix #170: "Determinism" annex; recommend sorted arrays

### LINKED ISSUE #170: we should recommend that all arrays be emitted in sorted order
to better support run-over-run analysis. not sure where this information goes.

### LINKED ISSUE #191: Write annex on determinism
Why determinism is important. Which SARIF properties to avoid -- or post-process away -- to ensure determinism.

### PR DESCRIPTION
**Pull request by ghost**
@michaelcfanning

### CODE REVIEW

File: document/AnnexE-Determinism [hunk 1]
```diff
@@ -0,0 +1,52 @@
...
+* `invocation.fileName` (because `fileName` is specified as being an absolute path, and
+tools might be stored in different directories on different machines)
+* `invocation.workingDirectory`
+* `invocation.environmentVariables`
+* The use of absolute file paths in `invocation.commandLine`
+(because builds performed on different machines might use a different root directory)
+* `annotatedCodeLocation.threadId`
+* `notification.threadId`
+* `notification.time`
+* `result.baselineState` (because it depends on a previous run.
```
Comment (michaelcfanning) [inline document/AnnexE-Determinism]: why is this a problem? if the baseline hasn't changed, the log files relation to it should be stable as well

File: document/AnnexE-Determinism [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+A result whose `baselineState` was `"new"` in a previous run would be `"existing"`
+in the next run.)
```
Comment (michaelcfanning) [inline document/AnnexE-Determinism]: right, but after that it would be completely stable (everything would stay as 'existing').

File: document/AnnexE-Determinism [hunk 3]
```diff
@@ -0,0 +1,52 @@
...
+* `run.id`
+* `run.automationId`
+* `run.baselineId`
```
Comment (michaelcfanning) [inline document/AnnexE-Determinism:62]: all these ids could be stable

File: document/AnnexE-Determinism [hunk 4]
```diff
@@ -0,0 +1,52 @@
...
+* `stackFrame.threadId`
+* `stackFrame.address`
+* The use of absolute file paths in URI-valued properties such as `physicalLocation.uri`
+* The presence of any non-deterministic elements in a property bag property
+
+Authors of analysis tools are encouraged to provide a mechanism
+(for example, a command line option such as `--deterministic`) which instructs the tool to
+omit non-deterministic elements from the log file.
```
Comment (michaelcfanning) [inline document/AnnexE-Determinism]: log processors could also simply skip over non-deterministic elements as described here as well.

File: document/AnnexE-Determinism [hunk 5]
```diff
@@ -0,0 +1,52 @@
+# Annex E (informative) Producing deterministic SARIF log files { - #Annex-E }
```
Comment (michaelcfanning) [inline document/AnnexE-Determinism:1]: you didn't mention ordering in this annex. results or other items which are emitted in a different order may be perceived as different by naive diffing mechanisms (such as source control systems).

File: document/AnnexE-Determinism (hunk 3 above)
Comment (ghost) [inline document/AnnexE-Determinism:62]: (I know you said "could be" stable; I'll address that at the end.)

A common example of `run.automationId` (this is the example given in the spec) would be the build id:

```
"automationId": "Build-14.0.1.2-20160518-15:48:02",
```

So if the same tool is run on the same inputs in two successive daily builds, the `automationId`s would differ.

As for `run.id`, the spec defines it to be

> a string which uniquely identifies the run

So it _has_ to change from run to run.

Now `run.baselineId` is defined to be

> a string which shall match the `id` property of some previous run

So if you run the tool twice against the same baseline, then `baselineId` will be the same. But in the common scenario where your baseline is yesterday's build, then `baselineId` will be different each time.

In short: I don't think either `run.automationId` or `run.id` could be stable. `run.baselineId` indeed could be stable in a particular scenario. But all in all, I think the guidance I currently give, which is that these elements "should not be used" if you want deterministic output, is good.

File: document/AnnexE-Determinism [hunk 6]
```diff
@@ -0,0 +1,77 @@
...
+In the case of SARIF, one could imagine a sequence of build steps where Steps A
+B, and C each run an analysis tool on a different set of targets, producing log files
+A.sarif, B.sarif, and C.sarif, and then build Step D performs an analysis on the aggregate
+of those log files. If the targets analyzed in Step B change
+but the targets analyzed in steps A and C do not, and if the contents of the
+SARIF log file are deterministic, then when the build is re-run, only Steps B and D
+need be performed.
+
+If it is desired to produce deterministic output, the following elements of the SARIF format
+should not be used. All of these elements are optional in the format:
```
Comment (michaelcfanning) [inline document/AnnexE-Determinism]: should we qualify this a bit? what if a build system is entirely consistent in the working directory? account name? what if it runs on a single machine? better would be language to note that these items hold potential for non-determinism, depending on infrastructure implementation. obviously things like startTime and processId are definitely or nearly always non-deterministic.

File: document/AnnexE-Determinism (hunk 3 above)
Comment (michaelcfanning) [inline document/AnnexE-Determinism:62]: I don't agree with this. i have a different understanding of 'uniquely' identifies which allows for run-over-run identity. again, think of an id like 'the SDL run for v1 of project Baz'. we are going to talk further on this. my previous position has been that we create an 'id' slot and describe ways that it might be used (either as a per-instance identifier or stable run-over-run). if we insist that an id serve only one of these functions, we need to provide slots for both purposes. because we do not, we can't exclude one possibility over the other.

---

In reply to: [66669250](https://github.com/sarif-standard/sarif-spec/pull/223#discussion_r66669250) [](ancestors = 66669250)

File: document/AnnexE-Determinism [hunk 7]
```diff
@@ -0,0 +1,77 @@
...
+
+In addition to avoiding non-deterministic elements of the file format,
+tools that wish to produce deterministic output must ensure that array and dictionary elements
+are produced in a predictable order.
+
+Some of the arrays defined in the SARIF format have a natural ordering
+for example, the `annotatedCodeLocation` objects within the `stack.Frames` property.
+Others, such as the array of `hash` objects within the `file.hashes` property, do not,
+but it would not be difficult for a tool to ensure that it always emitted them in the
+same order.
```
Comment (michaelcfanning) [inline document/AnnexE-Determinism]: yes, typically by explicitly sorting. in practice many unsorted constructs may be stable. the hashes are a good example. the problem is that it is not obvious to developers maintaining the code that the hash-emitting calls relate to preserving log file determinism. you can make this explicit in code comments (// hey, if you emit a new hash kind, please be sure to emit it after the current algos, in order to preserve log file ordering). better is to explicitly sort (and also provide a comment).

File: document/AnnexE-Determinism [hunk 8]
```diff
@@ -0,0 +1,77 @@
...
+
+The array of `result` objects presents more of a problem.
+A multi-threaded analysis tool analyzing multiple files in parallel might produce results
+in any order, and there is no natural order for the results.
+A tool might choose to order them, for example, first alphabetically by analysis target URI,
```
Comment (michaelcfanning) [inline document/AnnexE-Determinism:91]: a great sort order, thanks for articulating this

File: document/AnnexE-Determinism [hunk 9]
```diff
@@ -0,0 +1,77 @@
...
+then numerically by line number, then by column number, then alphabetically by rule id.
+
+For dictionaries such as the `run.rules` object or the `run.files` object,
+a tool might order the property names alphabetically, using a locale-insensitive ordering.
```
Comment (michaelcfanning) [inline document/AnnexE-Determinism:95]: We need to also specifically note the issue of case-sensitivity for file names. on windows, C:\BAZ\a.txt may sort differently if emitted as C:\baz\a.txt. These file paths are equivalent on Windows but the differing casing will fail naive diff mechanisms. So, we should add a section to this appendix around how to normalize case-insensitive things so they are deterministic in the log.

File: document/AnnexE-Determinism (hunk 3 above)
Comment (ghost) [inline document/AnnexE-Determinism:62]: Scenarios such as "find all SDL runs for v1 of project Baz" can be addressed with properties and tags:

```
runs: [
  {
    ...
    "properties": {
      "project": "Baz",
      "version": "1",
      "tags": [ "SDL" ]
    }
  }
]
```

I think this is better than using an "id"-like property, since runs might be bucketed in a variety of ways ("all runs for project 'Baz'", "all SDL runs", etc.), so a single id-like property wouldn't suffice in general.

---

In reply to: [66832989](https://github.com/sarif-standard/sarif-spec/pull/223#discussion_r66832989) [](ancestors = 66832989,66669250)

Comment (ghost) [inline document/AnnexE-Determinism:62]: Per our conversation: We are going to add `run.stableId`. So the current text, which advises against using `run.id`, `run.automationId`, or `run.baselineId` (which, if present, corresponds to a previous `run.id`), is correct as is.

---

In reply to: [66870708](https://github.com/sarif-standard/sarif-spec/pull/223#discussion_r66870708) [](ancestors = 66870708,66832989,66669250)

File: document/AnnexE-Determinism (hunk 4 above)
Comment (ghost) [inline document/AnnexE-Determinism]: Added text for this (lines 72-73)

---

In reply to: [64923943](https://github.com/sarif-standard/sarif-spec/pull/223#discussion_r64923943) [](ancestors = 64923943)

File: document/AnnexE-Determinism [hunk 10]
```diff
@@ -0,0 +1,136 @@
...
+of those log files. If the targets analyzed in Step B change
+but the targets analyzed in steps A and C do not, and if the contents of the
+SARIF log file are deterministic, then when the build is re-run, only Steps B and D
+need be performed.
+
+Authors of analysis tools are encouraged to provide a mechanism
+(for example, a command line option such as `--deterministic`) which instructs the tool to
+produce deterministic output.
+
+There are three issues to consider when producing deterministic output:
```
Comment (michaelcfanning) [inline document/AnnexE-Determinism]: If you are building a comprehensive list, we should also discuss two others. First, the paths issues, where there are deterministic relative paths against a non-deterministic root. Second, we may need some special language around baselining. Baselines are designed for run-over-run analysis, and so we can't toss 'baselineState' out. In other words, baselineState is designed to drive a specific diffing scenario. In order for it that value to be deterministic, results mgmt. tools need to rewrite baselines before check-in (so that all 'new' issues are 'existing') or the system needs a sufficiently subtle diff in order to accommodate transition from 'new' to 'existing'.

File: document/AnnexE-Determinism [hunk 11]
```diff
@@ -0,0 +1,136 @@
...
+* `invocation.endTime`
+* `invocation.processId`
+* `invocation.machine`
+* `invocation.account`
+* `invocation.fileName` (because `fileName` is specified as being an absolute path, and
+tools might be stored in different directories on different machines)
+* `invocation.workingDirectory`
+* `invocation.environmentVariables`
+* The use of absolute file paths in `invocation.commandLine`
+(because builds performed on different machines might use a different root directory)
```
Comment (michaelcfanning) [inline document/AnnexE-Determinism:56]: as above, need to pull this out and generalize it because it impacts every physical location

File: document/AnnexE-Determinism (hunk 2 above)
Comment (michaelcfanning) [inline document/AnnexE-Determinism]: as above, baselineState's purpose in life is to drive diff, we can't just toss it out, rather need to discuss it specifically

---

In reply to: [64923673](https://github.com/sarif-standard/sarif-spec/pull/223#discussion_r64923673) [](ancestors = 64923673)

File: document/AnnexE-Determinism [hunk 12]
```diff
@@ -0,0 +1,136 @@
...
+* `notification.threadId`
+* `notification.time`
+* `result.baselineState` (because it depends on a previous run.
+A result whose `baselineState` was `"new"` in a previous run would be `"existing"`
+in the next run.)
+* `run.id`
+* `run.automationId`
+* `run.baselineId`
+* `stackFrame.threadId`
+* `stackFrame.address`
```
Comment (michaelcfanning) [inline document/AnnexE-Determinism]: are you going to mention why the address may differ? e.g., due to security features that randomize the base address of an executable.

File: document/AnnexE-Determinism [hunk 13]
```diff
@@ -0,0 +1,136 @@
...
+* The use of absolute file paths in URI-valued properties such as `physicalLocation.uri`
```
Comment (michaelcfanning) [inline document/AnnexE-Determinism]: ah, here it is. we should address this as a specific topic, i think, and discuss how to handle it, via uriBaseId

File: document/AnnexE-Determinism [hunk 14]
```diff
@@ -0,0 +1,136 @@
...
+
+```
+"uri": "file:///C:/Code/file1.c"
+```
+
+A tool might choose to ignore this issue, since this would happen rarely, and the only
+consequence of ignoring it is that a subsequent step in a build process might execute
+even though it does not strictly have to.
+
+To produce absolutely deterministic output, the tool might instead choose to pull all
```
Comment (michaelcfanning) [inline document/AnnexE-Determinism]: 'pull' to lowercase or uppercase? is that a common use of this term? never heard that if so. 'convert' may be more typical

File: document/AnnexE-Determinism [hunk 15]
```diff
@@ -0,0 +1,136 @@
...
+file paths to lowercase or to uppercase, using a locale-insensitive case conversion:
+
+```
+"uri": "file:///C:/code/file1.c"
+```
+
+~IsoNote
+
+On the .NET platform, the recommendation is to pull strings to uppercase
```
Comment (michaelcfanning) [inline document/AnnexE-Determinism]: let's talk to Nick. what I'd say here is that, if string values are involved in security-sensitive comparisons, the recommendation is to convert to uppercase. if there is no special security significance to the literal files names within a log, lowercase is just fine. based on this, we can loosen the recommendation here. for our own build infrastructure, for example, i wouldn't see any security problems that might relate to converting to lowercase (because we don't have any special knowledge of individual files in the build, neither are there any special comparisons against directory or file names)

File: document/AnnexE-Determinism [hunk 16]
```diff
@@ -0,0 +1,136 @@
...
+"uri": "file:///C:/CODE/FILE1.C"
+```
+~
+
+## Post-processing { #Annex-E-PostProcessing }
+
+Even if an analysis tool does not produce deterministic output, a post-processing tool could
+produce deterministic output by removing or ignoring non-deterministic elements,
+reordering array elements and object properties,
+and adjusting the casing of case-insensitive string values
```
Comment (michaelcfanning) [inline document/AnnexE-Determinism]: more typically, you'd perform case-insensitive comparisons, rather than adjust the casing in a post-processing step. also, you need a period on the end of this sentence.

File: document/AnnexE-Determinism (hunk 6 above)
Comment (ghost) [inline document/AnnexE-Determinism]: Added a qualifying paragraph.

---

In reply to: [66832048](https://github.com/sarif-standard/sarif-spec/pull/223#discussion_r66832048) [](ancestors = 66832048)

File: document/AnnexE-Determinism (hunk 11 above)
Comment (ghost) [inline document/AnnexE-Determinism:56]: Done.

---

In reply to: [66999855](https://github.com/sarif-standard/sarif-spec/pull/223#discussion_r66999855) [](ancestors = 66999855)

File: document/AnnexE-Determinism (hunk 14 above)
Comment (ghost) [inline document/AnnexE-Determinism]: Changed to "convert"

---

In reply to: [67000883](https://github.com/sarif-standard/sarif-spec/pull/223#discussion_r67000883) [](ancestors = 67000883)

File: document/AnnexE-Determinism (hunk 16 above)
Comment (ghost) [inline document/AnnexE-Determinism]: There are two scenarios:
1. Log equality is determined by a naive content comparison or hash comparison. The guidance must be as I've given here. This is the Domino scenario.
2. Log equality is determined by an intelligent comparison. I'll add a mention of that possibility.

---

In reply to: [67001629](https://github.com/sarif-standard/sarif-spec/pull/223#discussion_r67001629) [](ancestors = 67001629)

File: document/AnnexE-Determinism (hunk 12 above)
Comment (ghost) [inline document/AnnexE-Determinism]: Done

---

In reply to: [67000530](https://github.com/sarif-standard/sarif-spec/pull/223#discussion_r67000530) [](ancestors = 67000530)

File: document/AnnexE-Determinism [hunk 17]
```diff
@@ -0,0 +1,199 @@
...
+
+~IsoNote
+
+On the .NET platform, the recommendation is to convert strings to uppercase
+using the method `String.ToUpperInvariant`, which would result in:
+
+```
+"uri": "file:///C:/CODE/FILE1.C"
+```
+~
```
Comment (ghost) [inline document/AnnexE-Determinism:147]: Still need to talk to Nick about this.

Comment (michaelcfanning) [inline document/AnnexE-Determinism:147]: ok, let's open an issue. as per other comments, i'd prefer we put language in on the security concerns that drive this practice to allow people to make a good decision. i.e., we don't want to advise upper-casing as a general practice, i think. it would be unusual to require it.

---

In reply to: [67031992](https://github.com/sarif-standard/sarif-spec/pull/223#discussion_r67031992) [](ancestors = 67031992)

File: document/AnnexE-Determinism [hunk 18]
```diff
@@ -0,0 +1,199 @@
...
+produce deterministic output by creating a new file that omits non-deterministic elements,
+reorders array elements and object properties,
+removes file path prefixes and introduces `uriBaseId` properties,
+and adjusts the casing of case-insensitive string values.
+
+In the second scenario, a post-processing step could intelligently compare the newly produced log
+to the log from a previous build by ignoring non-deterministic elements,
+ensuring that arrays have the same elements regardless of order,
+ignoring file path prefixes,
+and performing a case-insensitive comparison.
```
Comment (michaelcfanning) [inline document/AnnexE-Determinism]: and 'performing case-insensitive comparisons where appropriate' ?

Comment (ghost) [inline document/AnnexE-Determinism]: Fixed.

File: document/AnnexE-Determinism (hunk 17 above)
Comment (ghost) [inline document/AnnexE-Determinism:147]: Filed #231, "Is there a reason to recommend upper-casing?"

File: document/AnnexE-Determinism [hunk 19]
```diff
@@ -0,0 +1,199 @@
...
+```
+The `result` object that first appeared in `log_20160615.sarif` still appears in `log_20160616.sarif`,
+but since it existed in the baseline, its `baselineState` will now be `"existing"`.
+
+The result is that even though none of the analysis target files have changed,
+the log file has changed.
+
+Strictly speaking, this does not violate determinism. After all, the baseline file has changed,
+and the baseline file is one of the inputs to the analysis.
+But from a practical standpoint, this is still a problem, because... why?
```
Comment (ghost) [inline document/AnnexE-Determinism]: @michaelcfanning It's this last question that I'd like you to comment on. Yes, the contents of the log file changed, even though the analysis targets did not change. But the baseline, which _is_ an input to the analysis, _did_ change. Why should I care if the new log file doesn't match what's in the cache (and so triggers downstream build steps which would have been skipped if the logs had actually been identical)? This extra work is only done in the rare case when the _only_ changes from build to build are "new-to-existing" transitions. If, the next night, there are no further changes, then those results will now show up as "existing" both in the current and the previous logs, and this time, the extra build steps will indeed be skipped.

I'm having trouble believing this problem (if it is one) is worth the effort we're throwing at it. I'd be ok with omitting this entire section.

Comment (michaelcfanning) [inline document/AnnexE-Determinism]: this is still a problem where 'non-intelligent' comparisons are in play, e.g., when a file is hashed or simply compared directly to a previous version. so, the existence of baseline state in a log file can only be processed in a deterministic way with some 'intelligent' processing.

Comment (ghost) [inline document/AnnexE-Determinism]: My point is: how much do we care if a "non-intelligent" comparison, such as a simple file hash, decides that two log files which differ only in "new-to-existing" transitions are in fact different? The result of this decision by the non-intelligent comparer is that in this particular case, some build steps will be executed that don't actually need to be. The next night, if nothing else changes, these results will be "existing" in both the new build and in the baseline, and the extra steps will indeed be skipped.

So how much work do we want to do to avoid this extra work on those rare nights when the _only_ changes are "new-to-existing" transitions? My proposal was "None".

Comment (ghost) [inline document/AnnexE-Determinism]: Or maybe I'm missing your point. Are you simply saying that we should _warn_ people that this will occasionally happen, if they both (a) use the baselining feature, and (b) use a build system that makes decisions base on a non-intelligent comparison? I would agree with that.

Comment (michaelcfanning) [inline document/AnnexE-Determinism]: That's right. We don't need a lot of detail here. We simply need to note that the baselineState needs to be handled specially. Either an intelligent comparison needs to be in play (where baseline new == 'existing'), the system requires an intelligent diff, or you need a rewriting step (or some other additional work) to update the baseline to accommodate new->existing transitions.

---

In reply to: [67940333](https://github.com/sarif-standard/sarif-spec/pull/223#discussion_r67940333) [](ancestors = 67940333)

Comment (ghost) [inline document/AnnexE-Determinism]: ... or, if you do none of the above, your build system will occasionally do some extra work.

Comment (michaelcfanning) [inline document/AnnexE-Determinism]: Yes, assuming that the perceived differences don't result in a break that requires human review.
